cmd/api/main.go

```diff
@@ -59,18 +54,30 @@ func getKubernetesClient() *kubernetes.Clientset {
 	return client
 }
 
-func NewServer(k8sClient kubernetes.Interface, controller routers.Controller, cache *cache.Cache, cacheExpirations *cacheExpirations) *Server {
+func NewServer(k8sClient kubernetes.Interface, controller routers.Controller, cache *cache.Cache, cacheExpirations *routers.CacheExpirations) *Server {
+
 	router := gin.Default()
 	router.Use(otelgin.Middleware("")) // Empty string so the library sets the proper server
-	router.Use(auth.Authentication(k8sClient.AuthenticationV1().TokenReviews(), cache, cacheExpirations.Authorized, cacheExpirations.Unauthorized))
-	router.Use(auth.RBACAuthorization(k8sClient.AuthorizationV1().SubjectAccessReviews(), cache, cacheExpirations.Authorized, cacheExpirations.Unauthorized))
-	// TODO - Probably want to use cache for the discovery client
-	// See https://pkg.go.dev/k8s.io/client-go/discovery/cached/disk#NewCachedDiscoveryClientForConfig
-	router.Use(discovery.GetAPIResource(k8sClient.Discovery().RESTClient()))
-	router.GET("/apis/:group/:version/:resourceType", controller.GetAllResources)
-	router.GET("/apis/:group/:version/namespaces/:namespace/:resourceType", controller.GetNamespacedResources)
-	router.GET("/api/:version/:resourceType", controller.GetAllCoreResources)
-	router.GET("/api/:version/namespaces/:namespace/:resourceType", controller.GetNamespacedCoreResources)
+
+	apiGroup := router.Group("/api")
+	apisGroup := router.Group("/apis")
+	groups := [...]*gin.RouterGroup{apisGroup, apiGroup}
+	for _, group := range groups {
+		group.Use(auth.Authentication(k8sClient.AuthenticationV1().TokenReviews(), cache, cacheExpirations.Authorized, cacheExpirations.Unauthorized))
+		group.Use(auth.RBACAuthorization(k8sClient.AuthorizationV1().SubjectAccessReviews(), cache, cacheExpirations.Authorized, cacheExpirations.Unauthorized))
+		// TODO - Probably want to use cache for the discovery client
+		// See https://pkg.go.dev/k8s.io/client-go/discovery/cached/disk#NewCachedDiscoveryClientForConfig
+		group.Use(discovery.GetAPIResource(k8sClient.Discovery().RESTClient()))
+	}
+
+	router.GET("/livez", controller.Livez)
+	router.GET("/readyz", controller.Readyz)
+
+	apisGroup.GET("/:group/:version/:resourceType", controller.GetAllResources)
+	apisGroup.GET("/:group/:version/namespaces/:namespace/:resourceType", controller.GetNamespacedResources)
+
+	apiGroup.GET("/:version/:resourceType", controller.GetAllCoreResources)
+	apiGroup.GET("/:version/namespaces/:namespace/:resourceType", controller.GetNamespacedCoreResources)
```

Comment:
Purely visual/story-telling suggestion, feel free to ignore. I'm not sure GET and Use can be flipped:

```suggestion
	router.GET("/livez", controller.Livez)
	router.GET("/readyz", controller.Readyz)

	coreGroup := router.Group("/api")
	coreGroup.GET("/:version/:resourceType", controller.GetAllCoreResources)
	coreGroup.GET("/:version/namespaces/:namespace/:resourceType", controller.GetNamespacedCoreResources)

	nonCoreCoreGroup := router.Group("/apis")
	nonCoreCoreGroup.GET("/:group/:version/:resourceType", controller.GetAllResources)
	nonCoreCoreGroup.GET("/:group/:version/namespaces/:namespace/:resourceType", controller.GetNamespacedResources)

    // setup middleware for resource groups
	for _, group := range [...]*gin.RouterGroup{nonCoreCoreGroup, coreGroup} {
		group.Use(auth.Authentication(k8sClient.AuthenticationV1().TokenReviews(), cache, cacheExpirations.Authorized, cacheExpirations.Unauthorized))
		group.Use(auth.RBACAuthorization(k8sClient.AuthorizationV1().SubjectAccessReviews(), cache, cacheExpirations.Authorized, cacheExpirations.Unauthorized))
		// TODO - Probably want to use cache for the discovery client
		// See https://pkg.go.dev/k8s.io/client-go/discovery/cached/disk#NewCachedDiscoveryClientForConfig
		group.Use(discovery.GetAPIResource(k8sClient.Discovery().RESTClient()))
	}
```

Reply:
I will try it and let you know in this thread

Reply:
Just tried it out. The GET and Use cannot be flipped, that way the middleware don't get applied to the groups.